scene/main/canvas_item.cpp

```diff
@@ -784,40 +784,49 @@ void CanvasItem::draw_rect(const Rect2 &p_rect, const Color &p_color, bool p_fil
 		}
 
 		RenderingServer::get_singleton()->canvas_item_add_rect(canvas_item, p_rect, p_color);
+	} else if (p_width >= p_rect.size.width || p_width >= p_rect.size.height) {
+		Size2 pos = p_rect.position - Size2(0.5 * p_width, 0.5 * p_width);
+		Size2 size = p_rect.size + Size2(p_width, p_width);
+		RenderingServer::get_singleton()->canvas_item_add_rect(canvas_item, Rect2(pos, size), p_color);
 	} else {
-		// Thick lines are offset depending on their width to avoid partial overlapping.
-		// Thin lines don't require an offset, so don't apply one in this case
-		real_t offset;
-		if (p_width >= 2) {
-			offset = p_width / 2.0;
-		} else {
-			offset = 0.0;
+		Size2 pos = p_rect.position;
+		Size2 size = p_rect.size;
+		int width = (int)Math::round(p_width);
+		real_t offset = 0.5 * width;
```

Comment:
Nitpicking, but we can slightly optimize this. Multiplying an integer by 0.5 or by 0.5f should have the same exact behavior as long as this integer is less than 16777216, so we don't need the extra precision.

```suggestion
		real_t offset = 0.5f * width;
```

Reply:
Done.